TicketSearch.ipynb

```diff
@@ -0,0 +1,294 @@
+    "- `n_target_classes`: the number of unique ticket IDs.\n",
+    "- `integer_target`: this is True since our ticket IDs are integers between 0 and `n_target_classes - 1`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from thirdai import bolt\n",
+    "\n",
+    "model = bolt.UniversalDeepTransformer(\n",
+    "    data_types={\n",
+    "        \"query\": bolt.types.text(),\n",
+    "        \"id\": bolt.types.categorical(),\n",
+    "    },\n",
+    "    target=\"id\",\n",
+    "    n_target_classes=100_000,\n",
+    "    integer_target=True,\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# **Step 3: Cold-start the model**\n",
+    "It can be difficult to collect a supervised dataset that maps customer queries to relevant past ticket IDs (it would look something like the following)\n",
+    "```\n",
+    "query,id\n",
+    "npm install stuck on ubuntu machine,32887\n",
+    "...\n",
+    "```\n",
+    "So we can instead cold-start the model on our existing support ticket history. To do so, we need to give the model the path to the dataset as well as how to use the different columns of the dataset. \n",
+    "\n",
+    "Each support ticket has a title, a question body, an answer body, and tags. Since the title usually summarizes the problem well, we’ll call that a **strong column**. But the question body and the answer have all the details, so we’ll add them as **weak columns**, along with tags.\n",
+    "\n",
+    "You should feel free to play around and see what configuration gives you the best results for your dataset!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model.cold_start(\n",
+    "    filename=\"stackoverflow-train.csv\",\n",
+    "    strong_column_names=[\"title\"],\n",
+    "    weak_column_names=[\"question_body\", \"answer_body\", \"tags\"],\n",
+    "    learning_rate=0.001,\n",
+    "    epochs=5,\n",
+    "    metrics=[\"categorical_accuracy\"],\n",
+    ")\n",
+    "\n",
+    "# Save the model for later use.\n",
+    "model.save(\"ticketsearch.bolt\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# **Step 4: Display the results**\n",
+    "When we call model.predict() with our query, we get scores for each ticket ID. Let’s write a helper function to display this information to customers in a nice way."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df = pd.read_csv(\"stackoverflow-display.csv\")\n",
+    "\n",
+    "def search(query):\n",
+    "    print(\"Question:\", query)\n",
+    "    print(\"Results:\\n\")\n",
+    "\n",
+    "    scores = model.predict({\"query\": query})\n",
+    "    k = 3\n",
```

Comment:
It may be slightly more readable to make `k` a function argument with a default value of 3.